base/utfconvert.jl

```diff
@@ -0,0 +1,444 @@
+    flags == 0 && @inbounds return UTF8String(copy!(Vector{UInt8}(len), 1, dat, 1, len))
+    return encode_to_utf8(UInt32, dat, len + num2byte + num3byte*2 + num4byte*3)
+end
+
+#=
+@doc """
+@brief      Converts a UTF32String to a UTF8String
+
+@param[in]  ::Type{UTF8String}
+@param[in]  str::UTF32String
+
+@return     ::UTF8String
+@throws     ArgumentError
+""" ->
+=#
+function convert(::Type{UTF8String},  str::UTF32String)
+    dat = reinterpret(UInt32, str.data)
+    len = sizeof(dat) >>> 2
+    # handle zero length string quickly
+    len <= 1 && return UTF8String("")
+    # get number of bytes to allocate
+    len, flags, num4byte, num3byte, num2byte = check_string_utf32(dat, len-1)
+    flags == 0 && @inbounds return UTF8String(copy!(Vector{UInt8}(len), 1, dat, 1, len))
+    return encode_to_utf8(UInt32, dat, len + num2byte + num3byte*2 + num4byte*3)
+end
+
+#=
+@doc """
+@brief      Converts an already validated vector of UInt16 or UInt32 to a UTF8String
+
+@param[in]  T           type (UInt16 or UInt32)
+@param[in]  dat         Vector{T}
+@param[in]  len         length of output in bytes
+
+@return     ::UTF8String
+""" ->
+=#
+function encode_to_utf8{T<:Union(UInt16, UInt32)}(::Type{T}, dat, len)
+    buf = Vector{UInt8}(len)
+    out = 0
+    pos = 0
+    @inbounds while out < len
+        ch::UInt32 = dat[pos += 1]
+        # Handle ASCII characters
+        if ch <= 0x7f
+            buf[out += 1] = ch
+        # Handle 0x80-0x7ff
+        elseif ch < 0x800
+            buf[out += 1] = 0xc0 | (ch >>> 6)
+            buf[out += 1] = 0x80 | (ch & 0x3f)
+        # Handle 0x10000-0x10ffff (if input is UInt32)
+        elseif T == UInt32 && ch > 0xffff
+            output_utf8_4(buf, out, ch)
+            out += 4
+        # Handle surrogate pairs
+        elseif is_surrogate_codeunit(ch)
+            output_utf8_4(buf, out, get_supplementary(ch, dat[pos += 1]))
+            out += 4
+        # Handle 0x800-0xd7ff, 0xe000-0xffff UCS-2 characters
+        else
+            buf[out += 1] = 0xe0 | ((ch >>> 12) & 0x3f)
+            buf[out += 1] = 0x80 | ((ch >>> 6) & 0x3f)
+            buf[out += 1] = 0x80 | (ch & 0x3f)
+        end
+    end
+    UTF8String(buf)
+end
+
+#=
+"""
+@brief      Converts a UTF8String to a UTF32String
+
+@param[in]  ::Type{UTF32String}
+@param[in]  str::UTF8String
+
+@return     ::UTF32String
+@throws     ArgumentError
+"""
+=#
+function convert(::Type{UTF32String}, str::UTF8String)
+    dat = str.data
+    # handle zero length string quickly
+    sizeof(dat) == 0 && return empty_utf32
+    # Validate UTF-8 encoding, and get number of words to create
+    len, flags = check_string_utf8(dat)
+    # Optimize case where no characters > 0x7f
+    totlen = len+1
+    flags == 0 && return fast_utf_copy(Char, totlen, dat)
+    # has multi-byte UTF-8 sequences
+    buf = Vector{Char}(totlen)
+    @inbounds buf[totlen] = 0 # NULL termination
```

Comment:
It would be useful to remove any inbounds annotations  that have no measurable impact on performance (I suspect many of them make no difference in the actual runtime).